File: backend/pkg/gateways/database/repositories/extraction_result_repository.go
package repositories

import (
	"github.com/ijufumi/google-vision-sample/pkg/gateways/database/entities"
	"github.com/pkg/errors"
	"gorm.io/gorm"
)

type ExtractionResultRepository interface {
	GetAll(db *gorm.DB) ([]entities.ExtractionResult, error)
	GetByID(db *gorm.DB, id string) (*entities.ExtractionResult, error)
	Create(db *gorm.DB, entity entities.ExtractionResult) error
	Update(db *gorm.DB, entity entities.ExtractionResult) error
	Delete(db *gorm.DB, id string) error
}

func NewExtractionResultRepository() ExtractionResultRepository {
	return &extractionResultRepository{}
}

type extractionResultRepository struct {
}

func (r *extractionResultRepository) GetAll(db *gorm.DB) ([]entities.ExtractionResult, error) {
	var results []entities.ExtractionResult
	if err := db.Preload("ExtractedTexts").Find(&results).Error; err != nil {
		return nil, errors.Wrap(err, "ExtractionResultRepository#GetAll")
	}
	return results, nil
}

func (r *extractionResultRepository) GetByID(db *gorm.DB, id string) (*entities.ExtractionResult, error) {
	var result *entities.ExtractionResult
	if err := db.Preload("ExtractedTexts").Where(map[string]string{
		"id": id,
	}).First(result).Error; err != nil {
		return nil, errors.Wrap(err, "ExtractionResultRepository#GetByID")
	}
	return result, nil
}

func (r *extractionResultRepository) Create(db *gorm.DB, entity entities.ExtractionResult) error {
	if err := db.Create(&entity).Error; err != nil {
		return errors.Wrap(err, "ExtractionResultRepository#Create")
	}
	return nil
}

func (r *extractionResultRepository) Update(db *gorm.DB, entity entities.ExtractionResult) error {
	if err := db.Save(&entity).Error; err != nil {
		return errors.Wrap(err, "ExtractionResultRepository#Update")
	}
	return nil
}

func (r *extractionResultRepository) Delete(db *gorm.DB, id string) error {
	if err := db.Model(&entities.ExtractionResult{}).Delete(map[string]string{
		"id": id,
	}).Error; err != nil {
		return errors.Wrap(err, "ExtractionResultRepository#Delete")
	}
	return nil
}

File: frontend/src/repositories/ExtractionResultRepository.ts
import BaseRepository from "./BaseRepository"
import ExtractionResult, { Props } from "../models/ExtractionResult"
import Status from "../models/Status"

export interface ExtractionResultRepository {
  getAll(): Promise<ExtractionResult[]>

  getById(args: { id: string }): Promise<ExtractionResult>

  create(args: { file: File }): Promise<Status>

  delete(args: { id: string }): Promise<Status>
}

export class ExtractionResultRepositoryImpl
  extends BaseRepository
  implements ExtractionResultRepository
{
  create = async (args: { file: File }) => {
    const formData = new FormData()
    formData.set("file", args.file)
    const result = await this._upload({ path: "", body: formData })
    return new Status(result)
  }

  getAll = async () => {
    const results = await this._get({ path: "" })
    return results.map((p: Props) => new ExtractionResult(p))
  }

  getById = async (args: { id: string }) => {
    const result = await this._get({ path: `/${args.id}` })
    return new ExtractionResult(result)
  }

  delete = async (args: { id: string }) => {
    const result = await this._delete({ path: `/${args.id}` })
    return new Status(result)
  }
}
